### ISSUE TITLE
Implement vif parsing for Justice League Heros

### ISSUE THREAD

**Issue by Gnates** [bigianb/bgda-explorer#7]
Sorry, I'm still relatively new to Github, so I don't know if comments in Closed issues receive notifications. 

After you made a pre-release (https://github.com/bigianb/bgda-explorer/releases/tag/0.0.1), I tested it out and reported the crash I experienced: https://github.com/bigianb/bgda-explorer/issues/6

I believe you made some fixes after the crash report, but I don't know how to compile it to test it.

**Comment by bigianb** [comment]
The crash you were seeing was because you were using an Xbox version of the game. You need the ps2 game.
The only fixes were to make the crash less terminal. I tested with the ps2 version and the models show.

**Comment by Gnates** [comment]
> The crash you were seeing was because you were using an Xbox version of the game. You need the ps2 game. The only fixes were to make the crash less terminal. I tested with the ps2 version and the models show.

Oh I see now! I will go grab the PS2 version.
Thank you again for providing support for the game. 
Does this mean I can get these models into a 3D software like Blender?

**Comment by bigianb** [comment]
I just make a 0.0.2 release which shows the huntress. It may work with others but there are problems with reading the texture files so some will have issues.

**Comment by Gnates** [comment]
I just did an initial test with version 0.0.2 with Lutris. 

When I clicked play, I got the following error: 
**RUNDLL32.EXE This application could not be started.**

For 1-2 minutes I was google searching the error, and then the program window randomly appeared.

I tested the following:
```
Huntress LMP
- huntress.tex [preview works]
- huntress.vif [preview works]
- cape16.tex [preview works]
- cape16/vif [preview works]

KFrost LMP
- kfrost.tex [preview works]
- kfrost.vif [preview works]

Hawkgirl LMP
- hawkgirl.tex [preview works]
- hawgkgirl.vif [preview shows character head without body]
- mace.tex [preview works]
- mace.vif [preview works]

Zatanna LMP
- zatanna.tex [preview works]
- zatanna.vif [preview works]
- tophat.tex [crashes program]
- tophat.vif [crashes program]
- cape8.tex [preview works]
- cape8.vif [preview works]

Zat_IC LMP
- zat_ic.tex [preview works]
- zat_ic.vif [preview works]
- cape16.tex [preview works]
- cape16.vif [preview works]

Z_Classic LMP
- z_clasic.tex [preview works]
- z_clasic.vif [preview works]
- tophat.tex [crashes program]
- tophat.vif [crashes program]
- cape8.tex [preview works]
- cape8.vif [preview works]

WWoman LMP
- wwoman.tex [preview works]
- wwoman.vif [preview works]
- tiara.tex [preview works]
- tiara.vif [preview works]
- lasso.tex [preview works]
- lasso.vif [preview works]
- cape16.tex [preview works]
- cape16.vif [preview works]

WWToga LMP
- wwtoga.tex [preview works]
- wwtoga.vif [preview works]
- cape16.tex [preview works]
- cape16.vif [preview works]

WWBlack LMP
- wwblack.tex [preview works]
- wwblack.vif [preview works]

WWO_ARMD LMP
- wwo_armd.tex [preview works]
- wwo_armd.vif [preview works]
- tiara.tex [preview works]
- tiara.vif [preview works]
```
Attempting to export the Texture, Model or Posed Model will only export a PNG file with 0 bytes every time.

Are you experiencing the same thing in Windows?

**Comment by bigianb** [comment]
Thanks for testing. Some of the geometry is missing because the hack only looks for the first part of any mesh (most only have one part but some have more than one such as hawk). The crashes I think are due to the tex processing - the format is a bit different and I need to figure out what the difference is.
The anim does not work because the anm files are a different format ... I've not looked into that but I assume the difference will be small. The slider moves between animation frames but as the anm does not work then you want see anything (look at Baldurs Gate Dark Alliance to see it in action).

Finally, I'm not sure why the export does not work ... I'll take a look when I get some time.

**Comment by Gnates** [comment]
This time I launched Lutris from terminal and copy pasted the error that appeared after attempting to export a Model to OBJ:
```
Unhandled Exception: System.NotSupportedException: No imaging component suitable to complete this operation was found. ---> System.Runtime.InteropServices.COMException: Exception from HRESULT: 0x88982F50
   --- End of inner exception stack trace ---
   at System.Windows.Media.Imaging.BitmapEncoder.EnsureUnmanagedEncoder()
   at System.Windows.Media.Imaging.BitmapEncoder.Save(Stream stream)
   at WorldExplorer.DataExporters.VifExporter.WritePosedObj(String savePath, Model model, WriteableBitmap texture, AnimData pose, Int32 frame, Double scale)
   at WorldExplorer.MainWindow.Menu_Export_Model_Click(Object sender, RoutedEventArgs e)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseEvent(RoutedEventArgs e)
   at System.Windows.Controls.MenuItem.InvokeClickAfterRender(Object arg)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   at MS.Win32.UnsafeNativeMethods.DispatchMessage(MSG& msg)
   at System.Windows.Threading.Dispatcher.PushFrameImpl(DispatcherFrame frame)
   at System.Windows.Threading.Dispatcher.PushFrame(DispatcherFrame frame)
   at System.Windows.Application.RunDispatcher(Object ignore)
   at System.Windows.Application.RunInternal(Window window)
   at System.Windows.Application.Run(Window window)
   at WorldExplorer.App.Main()
wine: Unhandled exception 0xe0434352 in thread 9 at address 0x7b43d64d (thread 0009), starting debugger...
Unhandled exception: 0xe0434352 in 32-bit code (0x7b43d64d).
Register dump:
 CS:0023 SS:002b DS:002b ES:002b FS:0063 GS:006b
 EIP:7b43d64d ESP:0033e604 EBP:0033e678 EFLAGS:00200212(   - --  I   -A- - )
 EAX:7b42a439 EBX:7b640ff4 ECX:00000000 EDX:0033e624
 ESI:0033e6f4 EDI:0033e634
Stack dump:
0x0033e604:  006ebe4a 006ebe00 4f990bca e0434352
0x0033e614:  00000001 00000000 7b43d64d 00000005
0x0033e624:  80131515 00000000 00000000 00000000
0x0033e634:  00650000 00000000 0185ba04 00000005
0x0033e644:  00000001 00000000 00000000 00152ba0
0x0033e654:  0033e738 00000001 00000000 00000000
Backtrace:
=>0 0x7b43d64d in kernel32 (+0x1d64d) (0x0033e678)
  1 0x0077d135 EntryPoint+0xffffffff() in clr (0x0033e718)
  2 0x0077db8c EntryPoint+0xffffffff() in clr (0x0033e7e0)
  3 0x03e98379 in presentationcore.ni (+0x208378) (0x0033e82c)
  4 0x03e97f1d in presentationcore.ni (+0x207f1c) (0x0033e8a0)
  5 0x06a64029 (0x0033ecdc)
  6 0x06a63f3e (0x0033ed08)
  7 0x03e60c4e in presentationcore.ni (+0x1d0c4d) (0x0033ed1c)
  8 0x03e608be in presentationcore.ni (+0x1d08bd) (0x0033ed84)
  9 0x03e812f3 in presentationcore.ni (+0x1f12f2) (0x0033edc8)
  10 0x03e832d8 in presentationcore.ni (+0x1f32d7) (0x0033edd4)
  11 0x57148979 in presentationframework.ni (+0xb38978) (0x0033ede8)
  12 0x5859bf12 in windowsbase.ni (+0xbbf11) (0x0033ee00)
  13 0x5859be54 in windowsbase.ni (+0xbbe53) (0x0033ee3c)
  14 0x5859e422 in windowsbase.ni (+0xbe421) (0x0033ee84)
  15 0x5859e34c in windowsbase.ni (+0xbe34b) (0x0033eef0)
  16 0x79a96326 in mscorlib.ni (+0x376325) (0x0033ef04)
  17 0x79a962e1 in mscorlib.ni (+0x3762e0) (0x0033ef20)
  18 0x5859e23b in windowsbase.ni (+0xbe23a) (0x0033ef5c)
  19 0x5859a905 in windowsbase.ni (+0xba904) (0x0033ef9c)
  20 0x5859aafa in windowsbase.ni (+0xbaaf9) (0x0033efd8)
  21 0x5859c13b in windowsbase.ni (+0xbc13a) (0x0033f014)
  22 0x5859c08b in windowsbase.ni (+0xbc08a) (0x0033f034)
  23 0x5859bf12 in windowsbase.ni (+0xbbf11) (0x0033f04c)
  24 0x5859be54 in windowsbase.ni (+0xbbe53) (0x0033f088)
  25 0x5859a4f1 in windowsbase.ni (+0xba4f0) (0x0033f0dc)
  26 0x5859b9de in windowsbase.ni (+0xbb9dd) (0x0033f124)
  27 0x0039d08e (0x0033f158)
  28 0xf79aaacc in user32 (+0x9aacb) (0x0033f188)
  29 0xf79ab109 in user32 (+0x9b108) (0x0033f1d8)
  30 0xf79ad4c3 in user32 (+0x9d4c2) (0x0033f228)
  31 0xf796e7f0 in user32 (+0x5e7ef) (0x0033f338)
  32 0x585be59c in windowsbase.ni (+0xde59b) (0x0033f384)
  33 0x585ab2a2 in windowsbase.ni (+0xcb2a1) (0x0033f3d0)
  34 0x585aaf9e in windowsbase.ni (+0xcaf9d) (0x0033f3dc)
  35 0x5689122e in presentationframework.ni (+0x28122d) (0x0033f3ec)
  36 0x568910c4 in presentationframework.ni (+0x2810c3) (0x0033f40c)
  37 0x56890e8b in presentationframework.ni (+0x280e8a) (0x0033f41c)
  38 0x003f047c (0x0033f428)
  39 0x00651366 EntryPoint+0xffffffff() in clr (0x0033f434)
  40 0x00652f47 EntryPoint+0xffffffff() in clr (0x0033f488)
  41 0x00659d0a EntryPoint+0xffffffff() in clr (0x0033f4f8)
  42 0x0075ec57 EntryPoint+0xffffffff() in clr (0x0033f61c)
  43 0x0075ed4b EntryPoint+0xffffffff() in clr (0x0033f890)
  44 0x0075f373 EntryPoint+0xffffffff() in clr (0x0033fd98)
  45 0x0075f419 EntryPoint+0xffffffff() in clr (0x0033fdf0)
  46 0x0075eb9a EntryPoint+0xffffffff() in clr (0x0033fe30)
  47 0x0075af45 EntryPoint+0xffffffff() in clr (0x0033fe6c)
  48 0x1000cc2f EntryPoint+0xffffffff() in mscoreei (0x0033fea8)
  49 0x79007f16 EntryPoint+0x50c1() in mscoree (0x0033feb8)
  50 0x79004de3 EntryPoint+0x1f8e() in mscoree (0x0033fed8)
  51 0x7b4640c6 in kernel32 (+0x440c5) (0x0033ffd8)
  52 0x7b4622ce in kernel32 (+0x422cd) (0x0033ffec)
0x7b43d64d: addl	$12,%esp
Modules:
Module	Address			Debug info	Name (63 modules)
PE	  400000-  424000	Deferred        worldexplorer
PE	  650000-  cf8000	Export          clr
PE	  d00000-  df5000	Deferred        msvcr120_clr0400
PE	 32e0000- 3c8c000	Deferred        system.ni
PE	 3c90000- 4793000	Export          presentationcore.ni
PE	 47a0000- 493e000	Deferred        wpfgfx_v0400
PE	 4940000- 4a00000	Deferred        presentationnative_v0400
PE	 4a00000- 4a7d000	Deferred        clrjit
PE	 4a90000- 4aa3000	Deferred        nlssorting
PE	 7a50000- 7b24000	Deferred        diasymreader
PE	10000000-10078000	Export          mscoreei
PE	56580000-565ca000	Deferred        presentationframework.classic.ni
PE	56610000-57827000	Export          presentationframework.ni
PE	57a00000-57a07000	Deferred        presentationframework-systemxml.ni
PE	580d0000-5829c000	Deferred        system.xaml.ni
PE	58450000-5846b000	Deferred        uiautomationprovider.ni
PE	58480000-584af000	Deferred        uiautomationtypes.ni
PE	584e0000-588a5000	Export          windowsbase.ni
PE	60370000-60460000	Deferred        system.configuration.ni
PE	61a00000-62116000	Deferred        system.core.ni
PE	62160000-62876000	Deferred        system.xml.ni
PE	79000000-7904a000	Export          mscoree
PE	79720000-7a93b000	Export          mscorlib.ni
PE	7b420000-7b5d0000	Export          kernel32
PE	7bc10000-7bc14000	Deferred        ntdll
PE	f0c90000-f0d08000	Deferred        winmm
PE	f0ea0000-f0ea4000	Deferred        explorerframe
PE	f12f0000-f1341000	Deferred        comctl32
PE	f1430000-f14d5000	Deferred        comdlg32
PE	f1960000-f196b000	Deferred        winspool
PE	f19a0000-f19a3000	Deferred        usp10
PE	f29d0000-f29d9000	Deferred        msacm32
PE	f2c20000-f2c48000	Deferred        uiautomationcore
PE	f2dd0000-f2dd6000	Deferred        windowscodecs
PE	f3180000-f3a52000	Deferred        shell32
PE	f3ee0000-f3ee3000	Deferred        sxs
PE	f3f10000-f3f14000	Deferred        propsys
PE	f42f0000-f42f3000	Deferred        api-ms-win-core-winrt-string-l1-1-0
PE	f4310000-f4313000	Deferred        shcore
PE	f4330000-f4333000	Deferred        wtsapi32
PE	f4340000-f4344000	Deferred        uxtheme
PE	f6290000-f6294000	Deferred        opengl32
PE	f63c0000-f63c4000	Deferred        wined3d
PE	f66c0000-f66c4000	Deferred        d3d9
PE	f6700000-f6704000	Deferred        dwmapi
PE	f6730000-f6738000	Deferred        oleaut32
PE	f6840000-f6843000	Deferred        dwrite
PE	f6ea0000-f6ee2000	Deferred        crypt32
PE	f6f70000-f6f74000	Deferred        bcrypt
PE	f6f90000-f6f94000	Deferred        rsaenh
PE	f6fd0000-f6fd3000	Deferred        combase
PE	f6ff0000-f6ff3000	Deferred        api-ms-win-core-winrt-l1-1-0
PE	f7280000-f7284000	Deferred        winex11
PE	f7310000-f7313000	Deferred        kernelbase
PE	f7340000-f7343000	Deferred        api-ms-win-core-quirks-l1-1-0
PE	f7360000-f7364000	Deferred        rpcrt4
PE	f73f0000-f7418000	Deferred        ole32
PE	f7530000-f7534000	Deferred        imm32
PE	f77c0000-f77c4000	Deferred        version
PE	f77e0000-f77e7000	Deferred        gdi32
PE	f7910000-f79f2000	Export          user32
PE	f7b10000-f7b18000	Deferred        shlwapi
PE	f7b80000-f7b84000	Deferred        advapi32
Threads:
process  tid      prio (all id:s are in hex)
00000008 (D) C:\Program Files (x86)\WorldExplorer_002\Release\WorldExplorer.exe
	0000003a    0
	00000039    0
	00000038    0
	00000032    2
	00000031    0
	00000009    0 <==
0000000e services.exe
	0000002b    0
	00000026    0
	00000020    0
	00000010    0
	0000000f    0
0000001e winedevice.exe
	00000023    0
	00000022    0
	00000021    0
	0000001f    0
00000024 plugplay.exe
	00000028    0
	00000027    0
	00000025    0
00000029 winedevice.exe
	00000030    0
	0000002f    0
	0000002e    0
	0000002d    0
	0000002c    0
	0000002a    0
00000033 explorer.exe
	00000037    0
	00000036    0
	00000035    0
	00000034    0
System information:
    Wine build: wine-4.2
    Platform: i386 (WOW64)
    Version: Windows 7
```

**Comment by bigianb** [comment]
Just tested exporting the huntress model. Seems to work fine, so I guess your issue is linux specific.
![image](https://user-images.githubusercontent.com/2781371/59640834-304c4d80-9157-11e9-8fbc-27ee545bf2ff.png)

**Comment by Gnates** [comment]
> "Just tested exporting the huntress model. Seems to work fine, so I guess your issue is linux specific."

I believe you're right. 
I only installed .NetFramework4.6 with Lutris, so there maybe other components I need to install.

I have someone who is testing the program in Microsoft Windows. 
He said the following:

> "I managed to export most of them, seems can only export to obj, and even though we can see the skeleton data, we cant export it.
> Also, some meshes dont export, specifically capes and hair parts of the back, guess they are maybe like physics type meshes?"

> "I did notice that some them have their UVs flipped in export"

> "So far most of the models have loaded, 1 Green Lantern hasnt, 2 of the Flash models only load a tongue I think it is, none of Superman or Batman's capes export properly.
> Green Lanterns ring constructs dont load either.
> 
> I'll make a more comprehensive list later."

**Comment by 0Curtain0** [comment]
@bigianb 
Hey, I noticed you are still doing updates for the program. 
Any chance for another release I can test? 

Also, have you had any luck with viewing/extracting Zatanna's hair and hat (I think it was called tophat or similar) or character capes? If not, you know of any alternative tools for extracting those meshes? 
- https://gfycat.com/immediatedopeyaegeancat 
- https://gfycat.com/metallicuglyaustraliancattledog-justice-league-heroes-playstation-2

**Comment by bigianb** [comment]
Nothing material has changed ... I main added script decoding. I'm working getting the mini-map textures to export at the moment. I've not looked at the JLH files for a while ... concentrating on the Dark Alliance ones. I'll get to it, but can't say when.

**Comment by bigianb** [comment]
The tophat issue is that the texture is 4 bit per pixel and I don't support the method it uses for transfer in the c# code. Interestingly I recently added minimap support in the java code which should support this ... so I know how to do it. At this stage I'm more likely to add a java based export tool than to fix the c# code though.

**Comment by bigianb** [comment]
Wife was watching shakespeare on the TV so decided to make the c# port. New release uploaded: https://github.com/bigianb/bgda-explorer/releases/tag/0.0.3

**Comment by 0Curtain0** [comment]
Interesting, I hope it makes things easier. 
I'm hoping to bring more of these old character models back to life: https://gfycat.com/hospitableuglyaztecant 
I might have some questions in the future about the physics based clothing/capes from this game, cause it looks like I may have to manually adjust them to the characters.

Funny, with the current health crisis, I started slowly learning C++ and python. 
I wish I can contribute more in some way. With amazing tools like Noesis, UE Viewer, BGDA Explorer, Blender, etc. I'm wondering if there is a safe bet on which programming languages I should start with if I want to contribute to these sort of projects in the future. 

Looks like 0.0.3 was released a few seconds ago, I will make sure to provide feedback. Thank you.

**Comment by bigianb** [comment]
> I'm wondering if there is a safe bet on which programming languages I should start with if I want to contribute to these sort of projects in the future.

There's no safe bet ... but python is a great language to start with. C++ is just painful as a beginner.

**Comment by bigianb** [comment]
@0Curtain0 I just released 0.0.4 which fixes the problem with zatanna missing her hair. Maybe fixes other things too.
